### ISSUE TITLE
ICE when compiling non-virtual thunk of variadic function with darwinpcs ABI

### ISSUE THREAD

**Issue by lephilousophe** [llvm/llvm-project#54025]
When using multiple inheritance, a thunk is generated by LLVM to adjust the this pointer. In the thunk, the function call to the real function is made musttail.
However, when using darwinpcs ABI, LLVM cannot simplify the code call to make it tail recursive.

This results in this error which can be reproduced on multiple LLVM versions starting from LLVM 10 (LLVM 9 seems to work correctly and produce expected assembly).

This seems to be related to the fact that LLVM tries to push mandatory arguments on stack with this ABI instead of putting them in registers as it's done by Apple clang.

Here is a godbolt link: https://godbolt.org/z/bbvKG91zh

Here are the files:
[failing-889b42.cpp.txt](https://github.com/llvm/llvm-project/files/8126873/failing-889b42.cpp.txt)
[failing-889b42.sh.txt](https://github.com/llvm/llvm-project/files/8126874/failing-889b42.sh.txt)
[bugpoint-reduced-simplified.bc.txt](https://github.com/llvm/llvm-project/files/8126881/bugpoint-reduced-simplified.bc.txt)

Here is the compilation output:
```
$ clang++-15 -target aarch64-apple-darwin11 -stdlib=libc++ -std=c++11 -pedantic -O2 -c -o /tmp/failing.o /tmp/failing.cpp -v 
Ubuntu clang version 15.0.0-++20220222114410+3c0096a1d456-1~exp1~20220222114508.308
Target: aarch64-apple-darwin11
Thread model: posix
InstalledDir: /usr/bin
ignoring nonexistent directory "/usr/bin/../include/c++/v1"
ignoring nonexistent directory "/usr/include/c++/v1"
 (in-process)
 "/usr/lib/llvm-15/bin/clang" -cc1 -triple arm64-apple-ios5.0.0 -Wundef-prefix=TARGET_OS_ -Werror=undef-prefix -Wdeprecated-objc-isa-usage -Werror=deprecated-objc-isa-usage -Werror=implicit-function-declaration -emit-obj --mrelax-relocations -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name failing.cpp -mrelocation-model pic -pic-level 2 -mframe-pointer=non-leaf -ffp-contract=on -fno-rounding-math -funwind-tables=2 -faligned-alloc-unavailable -fcompatibility-qualified-id-block-type-checking -fvisibility-inlines-hidden-static-local-var -target-cpu apple-a7 -target-feature +v8a -target-feature +fp-armv8 -target-feature +neon -target-feature +crypto -target-feature +zcm -target-feature +zcz -target-feature +sha2 -target-feature +aes -target-abi darwinpcs -fallow-half-arguments-and-returns -mllvm -treat-scalable-fixed-error-as-warning -debugger-tuning=lldb -v -fcoverage-compilation-dir=/tmp -resource-dir /usr/lib/llvm-15/lib/clang/15.0.0 -stdlib=libc++ -internal-isystem /usr/local/include -internal-isystem /usr/lib/llvm-15/lib/clang/15.0.0/include -internal-externc-isystem /usr/include -O2 -pedantic -std=c++11 -fdeprecated-macro -fdebug-compilation-dir=/tmp -ferror-limit 19 -stack-protector 1 -fblocks -fencode-extended-block-signature -fregister-global-dtors-with-atexit -fgnuc-version=4.2.1 -fcxx-exceptions -fexceptions -fmax-type-align=16 -fcolor-diagnostics -vectorize-loops -vectorize-slp -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o /tmp/failing.o -x c++ /tmp/failing.cpp
clang -cc1 version 15.0.0 based upon LLVM 15.0.0 default target x86_64-pc-linux-gnu
ignoring nonexistent directory "/System/Library/Frameworks"
ignoring nonexistent directory "/Library/Frameworks"
#include "..." search starts here:
#include <...> search starts here:
 /usr/local/include
 /usr/lib/llvm-15/lib/clang/15.0.0/include
 /usr/include
End of search list.
fatal error: error in backend: failed to perform tail call elimination on a call site marked musttail
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: clang++-15 -target aarch64-apple-darwin11 -stdlib=libc++ -std=c++11 -pedantic -O2 -c -o /tmp/failing.o /tmp/failing.cpp -v
1.      <eof> parser at end of file
2.      Code generation
3.      Running pass 'Function Pass Manager' on module '/tmp/failing.cpp'.
4.      Running pass 'AArch64 Instruction Selection' on function '@_ZThn16_N1C3addEPKcz'
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
/lib/x86_64-linux-gnu/libLLVM-15.so.1(_ZN4llvm3sys15PrintStackTraceERNS_11raw_ostreamEi+0x31)[0x7f9fe5751f81]
/lib/x86_64-linux-gnu/libLLVM-15.so.1(_ZN4llvm3sys17RunSignalHandlersEv+0xee)[0x7f9fe574fcce]
/lib/x86_64-linux-gnu/libLLVM-15.so.1(_ZN4llvm3sys15CleanupOnSignalEm+0xfb)[0x7f9fe575133b]
/lib/x86_64-linux-gnu/libLLVM-15.so.1(+0xd93faa)[0x7f9fe567dfaa]
/lib/x86_64-linux-gnu/libLLVM-15.so.1(+0xd93f4b)[0x7f9fe567df4b]
/lib/x86_64-linux-gnu/libLLVM-15.so.1(_ZN4llvm3sys7Process4ExitEib+0x27)[0x7f9fe574c267]
clang++-15[0x413702]
/lib/x86_64-linux-gnu/libLLVM-15.so.1(_ZN4llvm18report_fatal_errorERKNS_5TwineEb+0x121)[0x7f9fe568ca51]
/lib/x86_64-linux-gnu/libLLVM-15.so.1(+0xda2926)[0x7f9fe568c926]
/lib/x86_64-linux-gnu/libLLVM-15.so.1(+0x2a3cc73)[0x7f9fe7326c73]
/lib/x86_64-linux-gnu/libLLVM-15.so.1(_ZNK4llvm14TargetLowering11LowerCallToERNS0_16CallLoweringInfoE+0x196a)[0x7f9fe5efe9ba]
/lib/x86_64-linux-gnu/libLLVM-15.so.1(_ZN4llvm19SelectionDAGBuilder14lowerInvokableERNS_14TargetLowering16CallLoweringInfoEPKNS_10BasicBlockE+0x160)[0x7f9fe5f1bf00]
/lib/x86_64-linux-gnu/libLLVM-15.so.1(_ZN4llvm19SelectionDAGBuilder11LowerCallToERKNS_8CallBaseENS_7SDValueEbbPKNS_10BasicBlockE+0x99f)[0x7f9fe5f06c9f]
/lib/x86_64-linux-gnu/libLLVM-15.so.1(_ZN4llvm19SelectionDAGBuilder9visitCallERKNS_8CallInstE+0x500)[0x7f9fe5eef2f0]
/lib/x86_64-linux-gnu/libLLVM-15.so.1(_ZN4llvm19SelectionDAGBuilder5visitERKNS_11InstructionE+0x70)[0x7f9fe5ee29b0]
/lib/x86_64-linux-gnu/libLLVM-15.so.1(_ZN4llvm16SelectionDAGISel16SelectBasicBlockENS_14ilist_iteratorINS_12ilist_detail12node_optionsINS_11InstructionELb0ELb0EvEELb0ELb1EEES6_Rb+0x149)[0x7f9fe5f75199]
/lib/x86_64-linux-gnu/libLLVM-15.so.1(_ZN4llvm16SelectionDAGISel20SelectAllBasicBlocksERKNS_8FunctionE+0x1778)[0x7f9fe5f74c28]
/lib/x86_64-linux-gnu/libLLVM-15.so.1(_ZN4llvm16SelectionDAGISel20runOnMachineFunctionERNS_15MachineFunctionE+0x8c7)[0x7f9fe5f727c7]
/lib/x86_64-linux-gnu/libLLVM-15.so.1(_ZN4llvm19MachineFunctionPass13runOnFunctionERNS_8FunctionE+0x12d)[0x7f9fe5adcbed]
/lib/x86_64-linux-gnu/libLLVM-15.so.1(_ZN4llvm13FPPassManager13runOnFunctionERNS_8FunctionE+0x39f)[0x7f9fe588c9bf]
/lib/x86_64-linux-gnu/libLLVM-15.so.1(_ZN4llvm13FPPassManager11runOnModuleERNS_6ModuleE+0x33)[0x7f9fe5893f43]
/lib/x86_64-linux-gnu/libLLVM-15.so.1(_ZN4llvm6legacy15PassManagerImpl3runERNS_6ModuleE+0x936)[0x7f9fe588d556]
/lib/x86_64-linux-gnu/libclang-cpp.so.15(_ZN5clang17EmitBackendOutputERNS_17DiagnosticsEngineERKNS_19HeaderSearchOptionsERKNS_14CodeGenOptionsERKNS_13TargetOptionsERKNS_11LangOptionsEN4llvm9StringRefEPNSE_6ModuleENS_13BackendActionESt10unique_ptrINSE_17raw_pwrite_streamESt14default_deleteISK_EE+0x185b)[0x7f9fecbbbbdb]
/lib/x86_64-linux-gnu/libclang-cpp.so.15(+0x1bc37c1)[0x7f9fecf057c1]
/lib/x86_64-linux-gnu/libclang-cpp.so.15(_ZN5clang8ParseASTERNS_4SemaEbb+0x254)[0x7f9febd4fe44]
/lib/x86_64-linux-gnu/libclang-cpp.so.15(_ZN5clang13CodeGenAction13ExecuteActionEv+0xb1)[0x7f9fecf01ae1]
/lib/x86_64-linux-gnu/libclang-cpp.so.15(_ZN5clang14FrontendAction7ExecuteEv+0x67)[0x7f9fed8ad457]
/lib/x86_64-linux-gnu/libclang-cpp.so.15(_ZN5clang16CompilerInstance13ExecuteActionERNS_14FrontendActionE+0x336)[0x7f9fed8041b6]
/lib/x86_64-linux-gnu/libclang-cpp.so.15(_ZN5clang25ExecuteCompilerInvocationEPNS_16CompilerInstanceE+0x29a)[0x7f9fed92729a]
clang++-15(_Z8cc1_mainN4llvm8ArrayRefIPKcEES2_Pv+0x996)[0x4132c6]
clang++-15[0x41151b]
/lib/x86_64-linux-gnu/libclang-cpp.so.15(+0x2142d52)[0x7f9fed484d52]
/lib/x86_64-linux-gnu/libLLVM-15.so.1(_ZN4llvm20CrashRecoveryContext9RunSafelyENS_12function_refIFvvEEE+0xdd)[0x7f9fe567df2d]
/lib/x86_64-linux-gnu/libclang-cpp.so.15(_ZNK5clang6driver10CC1Command7ExecuteEN4llvm8ArrayRefINS2_8OptionalINS2_9StringRefEEEEEPNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEPb+0x140)[0x7f9fed484840]
/lib/x86_64-linux-gnu/libclang-cpp.so.15(_ZNK5clang6driver11Compilation14ExecuteCommandERKNS0_7CommandERPS3_+0x3f3)[0x7f9fed44c153]
/lib/x86_64-linux-gnu/libclang-cpp.so.15(_ZNK5clang6driver11Compilation11ExecuteJobsERKNS0_7JobListERN4llvm15SmallVectorImplISt4pairIiPKNS0_7CommandEEEE+0x8a)[0x7f9fed44c3da]
/lib/x86_64-linux-gnu/libclang-cpp.so.15(_ZN5clang6driver6Driver18ExecuteCompilationERNS0_11CompilationERN4llvm15SmallVectorImplISt4pairIiPKNS0_7CommandEEEE+0x1a7)[0x7f9fed465e87]
clang++-15(main+0x2816)[0x410f86]
/lib/x86_64-linux-gnu/libc.so.6(+0x2dfd0)[0x7f9fe43d8fd0]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x7d)[0x7f9fe43d907d]
clang++-15(_start+0x25)[0x40e415]
clang: error: clang frontend command failed with exit code 70 (use -v to see invocation)
Ubuntu clang version 15.0.0-++20220222114410+3c0096a1d456-1~exp1~20220222114508.308
Target: aarch64-apple-darwin11
Thread model: posix
InstalledDir: /usr/bin
ignoring nonexistent directory "/usr/bin/../include/c++/v1"
ignoring nonexistent directory "/usr/include/c++/v1"
```

**Comment by lephilousophe** [comment]
This has been fixed.

**Comment by EugeneZelenko** [comment]
Fixed in 26cd258420c774254cc48330b1f4d23d353baf05.
